graph/src/components/store.rs

```diff
@@ -393,19 +415,255 @@ pub trait Store: Send + Sync + 'static {
+        versions_after: Vec<SubgraphVersionSummary>,
+        node_id: Option<NodeId>,
+    ) -> Vec<EntityOperation> {
+        fn should_have_assignment(version: &SubgraphVersionSummary) -> bool {
+            version.pending || version.current
+        }
+
+        let assignments_before = versions_before
+            .into_iter()
+            .filter(should_have_assignment)
+            .map(|v| v.deployment_id)
+            .collect::<HashSet<_>>();
+        let assignments_after = versions_after
+            .into_iter()
+            .filter(should_have_assignment)
+            .map(|v| v.deployment_id)
+            .collect::<HashSet<_>>();
+        let removed_assignments = &assignments_before - &assignments_after;
+        let added_assignments = &assignments_after - &assignments_before;
+
+        let mut ops = vec![];
+
+        if !removed_assignments.is_empty() {
+            debug!(
+                logger,
+                "Removing subgraph node assignments for {} subgraph deployment ID(s) ({})",
+                removed_assignments.len(),
+                removed_assignments
+                    .iter()
+                    .map(ToString::to_string)
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            );
+        }
+        if !added_assignments.is_empty() {
+            debug!(
+                logger,
+                "Adding subgraph node assignments for {} subgraph deployment ID(s) ({})",
+                removed_assignments.len(),
+                removed_assignments
+                    .iter()
+                    .map(ToString::to_string)
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            );
+        }
+        if removed_assignments.is_empty() && added_assignments.is_empty() {
+            debug!(
+                logger,
+                "No subgraph node assignment additions/removals needed"
+            );
+        }
+
+        ops.extend(
+            removed_assignments
+                .into_iter()
+                .map(|deployment_id| EntityOperation::Remove {
+                    key: SubgraphDeploymentAssignmentEntity::key(deployment_id),
+                }),
+        );
+        ops.extend(added_assignments.iter().flat_map(|deployment_id| {
+            SubgraphDeploymentAssignmentEntity::new(
+                node_id
+                    .clone()
+                    .expect("Cannot create new subgraph deployment assignment without node ID"),
+            )
+            .write_operations(deployment_id)
+        }));
+
+        ops
+    }
 
     /// Check if the store is accepting queries for the specified subgraph.
     /// May return true even if the specified subgraph is not currently assigned to an indexing
     /// node, as the store will still accept queries.
-    fn is_deployed(&self, id: &SubgraphDeploymentId) -> Result<bool, Error>;
+    fn is_deployed(&self, id: &SubgraphDeploymentId) -> Result<bool, Error> {
```

Comment:
Note: this method is not new and has not actually changed -- I just moved the body to here from the diesel/postgres store